File: dcp-solutions/2024-12-22/solution.py
```python
from collections import defaultdict

def shortest_substring(s, chars):
    if not s or not chars:
        return None
    
    char_count = defaultdict(int, {char: 0 for char in chars})
    required_chars = len(char_count)
    
    left = start = 0
    min_length = float('inf')
    min_window = ""
    
    for right in range(len(s)):
        char = s[right]
        
        if char not in char_count or char_count[char] == 0:
            continue
            
        char_count[char] += 1
        
        if char_count[char] == 1:
            required_chars -= 1
            
        if required_chars == 0:
            while left <= right:
                min_length = min(min_length, right - left + 1)
                
                if s[left] not in char_count or char_count[s[left]] == 0:
                    break
                
                char_count[s[left]] -= 1
                if char_count[s[left]] == 0:
                    required_chars += 1
                
                left += 1
            
            start = left
    
    return min_window if min_length != float('inf') else None
```

File: dcp-solutions/2024-12-22/solution.php
```php
function shortestSubstring($s, $ chars) {
    if (empty($chars)) {
        return "";
    }

    $charCount = array_count_values(str_split($chars));
    $left = 0;
    $minLen = INF;
    $minStr = "";

    $charSet = [];

    foreach ($charCount as $char => $count) {
        $charSet[$char] = 0;
    }

    for ($right = 0; $right < strlen($s); $right++) {
        if (isset($charSet[$s[$right]])) {
            $charSet[$s[$right]]++;
            if ($charSet[$s[$right]] == $charCount[$s[$right]]) {
                unset($charSet[$s[$right]]);
            }
        }

        while (!empty($charSet) && count($charSet) == count($charCount)) {
            if ($right - $left + 1 < $minLen) {
                $minLen = $right - $left + 1;
                $minStr = substr($s, $left, $minLen);
            }

            if (isset($charSet[$s[$left]])) {
                $charSet[$s[$left]]--;
                if ($charSet[$s[$left]] == 0) {
                    unset($charSet[$s[$left]]);
                }
            }

            $left++;
        }
    }

    return $minStr ? $minStr : null;
}
```